File: MovieAppMVVMRetrofitJava/app/src/main/java/com/masterclass/movieappmvvmretrofitjava/activities/MainActivity.java
package com.masterclass.movieappmvvmretrofitjava.activities;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import android.annotation.SuppressLint;
import android.os.Bundle;

import com.masterclass.movieappmvvmretrofitjava.R;
import com.masterclass.movieappmvvmretrofitjava.adapters.MoviesAdapter;
import com.masterclass.movieappmvvmretrofitjava.databinding.ActivityMainBinding;
import com.masterclass.movieappmvvmretrofitjava.models.Movie;
import com.masterclass.movieappmvvmretrofitjava.viewModels.MainActivityViewModel;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    private ArrayList<Movie> movies;
    private RecyclerView recyclerView;
    private MoviesAdapter moviesAdapter;
    private SwipeRefreshLayout swipeRefreshLayout;
    private ActivityMainBinding binding;
    private MainActivityViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        viewModel = new ViewModelProvider(this).get(MainActivityViewModel.class);

        getPopularMovies();

        swipeRefreshLayout = binding.swipeLayout;
        swipeRefreshLayout.setColorSchemeResources(R.color.black);
        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                getPopularMovies();
            }
        });
    }

    private void getPopularMovies() {
        viewModel.getAllMovies().observe(this, new Observer<List<Movie>>() {
            @Override
            public void onChanged(List<Movie> moviesFromLiveData) {
                movies = (ArrayList<Movie>) moviesFromLiveData;
                displayMoviesInRecyclerView();
                if (swipeRefreshLayout.isRefreshing()){
                    swipeRefreshLayout.setRefreshing(false);
                }
            }
        });
    }

    @SuppressLint("NotifyDataSetChanged")
    private void displayMoviesInRecyclerView() {
        recyclerView = binding.recyclerView;
        moviesAdapter = new MoviesAdapter(this,movies);
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setLayoutManager(new GridLayoutManager(this,2));
        recyclerView.setAdapter(moviesAdapter);
        moviesAdapter.notifyDataSetChanged();
    }
}

File: MovieAppMVVMRetrofitKotlin/app/src/main/java/com/masterclass/movieappmvvmretrofitkotlin/activities/MainActivity.kt
package com.masterclass.movieappmvvmretrofitkotlin.activities

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.masterclass.movieappmvvmretrofitkotlin.R
import com.masterclass.movieappmvvmretrofitkotlin.adapters.MoviesAdapter
import com.masterclass.movieappmvvmretrofitkotlin.databinding.ActivityMainBinding
import com.masterclass.movieappmvvmretrofitkotlin.models.Movie
import com.masterclass.movieappmvvmretrofitkotlin.viewModels.MainActivityViewModel

class MainActivity : AppCompatActivity() {

    private lateinit var movies: ArrayList<Movie>
    private lateinit var recyclerView: RecyclerView
    private lateinit var moviesAdapter: MoviesAdapter
    private lateinit var swipeRefreshLayout: SwipeRefreshLayout
    private lateinit var binding: ActivityMainBinding
    private lateinit var viewModel: MainActivityViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        viewModel = ViewModelProvider(this).get(MainActivityViewModel::class.java)

        getPopularMovies()

        swipeRefreshLayout = binding.swipeLayout
        swipeRefreshLayout.setColorSchemeResources(R.color.black)
        swipeRefreshLayout.setOnRefreshListener {
            getPopularMovies()
        }
    }

    private fun getPopularMovies() {
        viewModel.getAllMovies().observe(this, Observer { moviesFromLiveData ->
            movies = moviesFromLiveData as ArrayList<Movie>
            displayMoviesInRecyclerView()
            if (swipeRefreshLayout.isRefreshing) {
                swipeRefreshLayout.isRefreshing = false
            }
        })
    }

    @Suppress("NotifyDataSetChanged")
    private fun displayMoviesInRecyclerView() {
        recyclerView = binding.recyclerView
        moviesAdapter = MoviesAdapter(this, movies)
        recyclerView.itemAnimator = DefaultItemAnimator()
        recyclerView.layoutManager = GridLayoutManager(this, 2)
        recyclerView.adapter = moviesAdapter
        moviesAdapter.notifyDataSetChanged()
    }
}